__tests__/Auth0Client.test.ts

```diff
@@ -1449,4 +1440,148 @@ describe('Auth0Client', () => {
       expect(decodedToken).toBeUndefined();
     });
   });
+
+  describe('getIdTokenClaims', () => {
+    describe('when loginWithRedirect', () => {
+      it('returns undefined if there is no cache', async () => {
+        const auth0 = setup();
+
+        jest.spyOn(auth0['cache'], 'get').mockReturnValueOnce(undefined);
+
+        const decodedToken = await auth0.getIdTokenClaims();
+        expect(decodedToken).toBeUndefined();
+      });
+
+      it('returns the ID token claims', async () => {
+        const auth0 = setup({ scope: 'foo' });
+        await loginWithRedirect(auth0);
+
+        expect(await auth0.getIdTokenClaims()).toHaveProperty('exp');
+        expect(await auth0.getIdTokenClaims()).not.toHaveProperty('me');
+        expect(await auth0.getIdTokenClaims({})).toHaveProperty('exp');
+        expect(
+          await auth0.getIdTokenClaims({ audience: 'default' })
+        ).toHaveProperty('exp');
+        expect(await auth0.getIdTokenClaims({ scope: 'foo' })).toHaveProperty(
+          'exp'
+        );
+        expect(
+          await auth0.getIdTokenClaims({ audience: 'invalid' })
+        ).toBeUndefined();
+      });
+
+      it('returns the ID token claims with custom scope', async () => {
+        const auth0 = setup({
+          scope: 'scope1',
+          advancedOptions: {
+            defaultScope: 'scope2'
+          }
+        });
+        await loginWithRedirect(auth0, { scope: 'scope3' });
+
+        expect(
+          await auth0.getIdTokenClaims({ scope: 'scope1 scope2 scope3' })
+        ).toHaveProperty('exp');
+      });
+
+      describe('when using refresh tokens', () => {
+        it('returns the ID token claims with offline_access', async () => {
+          const auth0 = setup({ scope: 'foo', useRefreshTokens: true });
+          await loginWithRedirect(auth0);
+
+          expect(
+            await auth0.getIdTokenClaims({ scope: 'foo offline_access' })
+          ).toHaveProperty('exp');
+        });
+
+        it('returns the ID token claims with custom scope and offline_access', async () => {
+          const auth0 = setup({
+            scope: 'scope1',
+            advancedOptions: {
+              defaultScope: 'scope2'
+            },
+            useRefreshTokens: true
+          });
+          await loginWithRedirect(auth0, { scope: 'scope3' });
+
+          expect(
+            await auth0.getIdTokenClaims({
+              scope: 'scope1 scope2 scope3 offline_access'
+            })
+          ).toHaveProperty('exp');
+        });
+      });
+    });
+
+    describe('when loginWithPopup', () => {
```

Comment:
@stevehobbsdev Here the tests are duplicated because the getIdTokenClaims, when used without mocking, needs a call to either `loginWithRedirect` or `loginWithPopup`. I feel like we want to ensure the getIdTokenClaims return the correct values in both cases.

Reply:
I feel like we can try the other suggestion here of putting `loginWithRedirect` and `loginWithPopup` into an array and generating tests that way, if the tests between the two of them are exactly the same. We can reduce the duplication and cover both scenarios.

My only worry was that it would be less readable but let's make sure we document in the code what we're doing, and why.

Reply:
I pushed a commit for this, let me know what u think.